src/main/java/org/icatproject/ids/dataSelection/DataSelectionBase.java

```diff
@@ -0,0 +1,265 @@
+package org.icatproject.ids.dataSelection;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.icatproject.IcatException_Exception;
+import org.icatproject.ids.enums.DeferredOp;
+import org.icatproject.ids.enums.RequestType;
+import org.icatproject.ids.exceptions.BadRequestException;
+import org.icatproject.ids.exceptions.DataNotOnlineException;
+import org.icatproject.ids.exceptions.InternalException;
+import org.icatproject.ids.exceptions.NotImplementedException;
+import org.icatproject.ids.models.DataFileInfo;
+import org.icatproject.ids.models.DataInfoBase;
+import org.icatproject.ids.services.ServiceProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class DataSelectionBase {
+
+    protected final static Logger logger = LoggerFactory.getLogger(DataSelectionBase.class);
+
+    protected SortedMap<Long, DataInfoBase> dsInfos;
+    protected SortedMap<Long, DataInfoBase> dfInfos;
+    protected Set<Long> emptyDatasets;
+    protected List<Long> invids;
+    protected List<Long> dsids;
+    protected List<Long> dfids;
+    protected RequestType requestType;
+
+
+    protected static ExecutorService threadPool;
+    static { threadPool = Executors.newCachedThreadPool(); }
+
+    private Map<String, PreparedStatus> preparedStatusMap = new ConcurrentHashMap<>();
+
+    class PreparedStatus {
+        public ReentrantLock lock = new ReentrantLock();
+        public Long fromElement;
+        public Future<?> future;
+
+    }
+
+    protected DataSelectionBase(SortedMap<Long, DataInfoBase> dsInfos, SortedMap<Long, DataInfoBase> dfInfos, Set<Long> emptyDatasets, List<Long> invids2, List<Long> dsids, List<Long> dfids, RequestType requestType) {
+
+        this.dsInfos = dsInfos;
+        this.dfInfos = dfInfos;
+        this.emptyDatasets = emptyDatasets;
+        this.invids = invids2;
+        this.dsids = dsids;
+        this.dfids = dfids;
+        this.requestType = requestType;
+    }
+
+    public abstract boolean isPrepared(String preparedId) throws InternalException;
+
+    /**
+     * To get the DataInfos that is currently worked with, depending on StorageUnit
+     * @return
+     */
+    public abstract SortedMap<Long, DataInfoBase> getPrimaryDataInfos();
+
+    public abstract boolean existsInMainStorage(DataInfoBase dataInfo) throws InternalException;
+
+    public abstract void queueDelete() throws NotImplementedException, InternalException;
+
+    public abstract void scheduleTasks(DeferredOp operation) throws NotImplementedException, InternalException;
```

Comment:
I'd like to implement this method here and use getPrimaryDataInfos(), instead of having it abstract. BUT In case of SingleLevelStorage we need a NotImplementedException here and this will not be the case, because getPrimaryDataInfos() will provide an empty Map in that case, and nothing will happen - not even an Exception. But if we change the corresponding tests, we could do it. But maybe we shouldn't change the behavior of the ids server.